### ISSUE TITLE
App Router: Support for removing default locale from URL

### ISSUE THREAD

**Issue by lachlanjc** [QuiiBz/next-international#62]
This library is excellent! I'd love to recreate the default functionality of the pages router, where the default locale isn't included in the URL, so e.g. "/about" is en-US, and "/en-GB/about" is en-GB or another locale. Is that possible with the middleware?

**Comment by QuiiBz** [comment]
Thanks! This isn't currently possible because we're using the locale parameter from the URL to detect which locale to use on both the server and client. It might be possible to support it using a combination of cookies and custom headers.

Just being curious, why do you prefer URLs without the default locale included in the URL?

**Comment by lachlanjc** [comment]
Yep, that makes sense. It's entirely cosmetic, but we want this for https://watershed.com (which uses next-international right now), since that's the setup we've had with the pages router. Since we only localize to UK English, and not all of our content is localized, it feels overbearing to make localization a part of every URL around the world when there's only one partial variant for one country.

**Comment by lachlanjc** [comment]
Since only the default locale wouldn't have a locale in the pathname, could you continue using the same URL param solution but return the default locale if there isn't one? Then use rewrites from e.g. `/blog` to `/en-US/blog`?

**Comment by QuiiBz** [comment]
Got it working (`/` working as `/en` or `/fr`) by adding a rewrite:

![Screenshot 2023-07-14 at 08 23 12](https://github.com/QuiiBz/next-international/assets/43268759/96cabf32-abfd-4755-b228-bd0ad0c16424)

And avoiding running the middleware on the root (note the `.*` which becomes `.+`):

![Screenshot 2023-07-14 at 08 22 35](https://github.com/QuiiBz/next-international/assets/43268759/f8f7bfd6-cc46-4ac4-98ea-66a72cec1ed5)

Let me know if this technique also works for you, and I'll update the README to mention it.

**Comment by QuiiBz** [comment]
Hi @lachlanjc, were you able to achieve what you wanted or do we still need to explore another way to support this use-case?

**Comment by ttiras** [comment]
hey @QuiiBz, i have seen this in a tutorial from a vercel insider and it removes all the locales from the url: 

```
import { NextRequest, NextResponse } from 'next/server'
import { match } from '@formatjs/intl-localematcher';
import Negotiator from 'negotiator';

let locales = ['en', 'es'];
export let defaultLocale = 'en';

function getLocale(request: Request): string {
  const headers = new Headers(request.headers)
  const acceptLanguage = headers.get("accept-language")
  if (acceptLanguage) {
    headers.set('accept-language', acceptLanguage.replaceAll("_", "-"))
  }

  const headersObject = Object.fromEntries(headers.entries());
  const languages = new Negotiator({ headers: headersObject }).languages();
  return match(languages, locales, defaultLocale);
}

export function middleware(request: NextRequest) {
  let locale = getLocale(request) ?? defaultLocale
  const pathname = request.nextUrl.pathname

  const newUrl = new URL(`/${locale}${pathname}`, request.nextUrl);

  // e.g. incoming request is /products
  // The new URL is now /en/products
  return NextResponse.rewrite(newUrl)
}

export const config = {
  matcher: [
    // Skip all internal paths (_next)
    '/((?!_next|api|favicon.ico).*)',
    // Optional: only run on root (/) URL
    // '/'
  ],
}
```

Hope that helps.

**Comment by EdmundKorley** [comment]
@ttiras I tried something similar to this but it can't work because of the current version's [use of `usePathname`] to get the local layout segment. `usePathname` doesn't return the 'rewritten' url path.

[use of `usePathname`]: https://github.com/QuiiBz/next-international/blob/main/packages/next-international/src/app/client/create-use-current-locale.ts

I ended up just forking the repo and replacing that call with [`useParams`] to push out a project.

[`useParams`]: https://github.com/EdmundKorley/next-international/commit/2ae0ac7be4b1555eb8098f0cba3e86e913838d4c

@QuiiBz what do you think about the approach in the commit?

**Comment by QuiiBz** [comment]
@EdmundKorley I think this is a totally acceptable change to land in next-international directly. Would you like to send a PR?

**Comment by ttiras** [comment]
Hey @EdmundKorley, thanks for trying and finding a solution.

@QuiiBz, I wanted to emphasize that this feature could be a highly valuable addition to the project, thank you.

**Comment by QuiiBz** [comment]
I mostly agree with what you said, @ttiras. To provide a bit more context, the `[locale]` param is used to retrieve the user's locale on the client. On the server, we no longer use `[locale]`, but instead an `X-Next-Locale` header or `Next-Locale` cookie that is set by the middleware. I believe we could refactor the client part to remove the dependency on `[locale]` and instead retrieve the user's locale using the `Next-Locale` cookie.

**Comment by EdmundKorley** [comment]
Thanks for the context @ttiras. I'll work on a PR to allows users to opt-in functionality that rewrites urls with no locale prefix to be prefixed with the default locale.

**Comment by QuiiBz** [comment]
I'm just waiting for [this PR](https://github.com/QuiiBz/next-international/pull/79) to land before triggering a new release, which will include #83

**Comment by QuiiBz** [comment]
Released in [0.8.0](https://github.com/QuiiBz/next-international/releases/tag/0.8.0)!
